source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,801 @@
+/*
+ * FreeRTOS V202012.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/*
+ * Demo for showing how to use the Device Shadow library's API. This version
+ * of the Device Shadow API provides macros and helper functions for assembling MQTT topics
+ * strings, and for determining whether an incoming MQTT message is related to the
+ * device shadow.
+ *
+ * This example assumes there is a powerOn state in the device shadow. It does the
+ * following operations:
+ * 1. Assemble strings for the MQTT topics of device shadow, by using macros defined by the Device Shadow library.
+ * 2. Subscribe to those MQTT topics using the MQTT Agent.
+ * 3. Register callbacks for incoming shadow topic publishes with the subsciption_manager.
+ * 3. Publish to report the current state of powerOn.
+ * 5. Check if powerOn has been changed and send an update if so.
+ * 6. If a publish to update reported state was sent, wait until either prvIncomingPublishUpdateAcceptedCallback
+ *    or prvIncomingPublishUpdateRejectedCallback handle the response.
+ * 7. Wait until time for next check and repeat from step 5.
+ *
+ * Meanwhile, when prvIncomingPublishUpdateDeltaCallback recives changes to shadow state, it will apply them on the
+ * device
```

Comment:
```suggestion
 * Meanwhile, when prvIncomingPublishUpdateDeltaCallback receives changes to the shadow state,
 * it will apply them on the device.
```